pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -935,3 +945,82 @@ func storePipelineSpec(ctx context.Context, pr *v1beta1.PipelineRun, ps *v1beta1
 	}
 	return nil
 }
+
+func (c *Reconciler) updatePipelineRunStatusFromInformer(pr *v1beta1.PipelineRun) error {
+	pipelineRunLabels := getTaskrunLabels(pr, "")
+	taskRuns, err := c.taskRunLister.TaskRuns(pr.Namespace).List(labels.SelectorFromSet(pipelineRunLabels))
+	if err != nil {
+		c.Logger.Errorf("Could not list TaskRuns %#v", err)
+		return err
+	}
+	// Store a list of Condition TaskRuns for each PipelineTask (by name)
+	conditionTaskRuns := make(map[string][]*v1beta1.TaskRun)
+	// Map PipelineTask names to TaskRun names that were already in the status
+	taskRunByPipelineTask := make(map[string]string)
+	// First loop over all the TaskRuns associated to Tasks
+	for _, taskrun := range taskRuns {
+		lbls := taskrun.GetLabels()
+		pipelineTaskName := lbls[pipeline.GroupName+pipeline.PipelineTaskLabelKey]
+		if _, ok := lbls[pipeline.GroupName+pipeline.ConditionCheckKey]; ok {
+			// Save condition for looping over them after this
+			if _, ok := conditionTaskRuns[pipelineTaskName]; !ok {
+				// If it's the first condition taskrun, initialise the slice
+				conditionTaskRuns[pipelineTaskName] = []*v1beta1.TaskRun{}
+			}
+			conditionTaskRuns[pipelineTaskName] = append(conditionTaskRuns[pipelineTaskName], taskrun)
+			continue
+		}
+		// Map pipeline task to taskrun name
+		taskRunByPipelineTask[pipelineTaskName] = taskrun.Name
+		if _, ok := pr.Status.TaskRuns[taskrun.Name]; !ok {
+			// This taskrun was missing from the status.
+			// Add it without conditions, which are handled in the next loop
+			pr.Status.TaskRuns[taskrun.Name] = &v1beta1.PipelineRunTaskRunStatus{
+				PipelineTaskName: pipelineTaskName,
+				Status:           &taskrun.Status,
+				ConditionChecks:  nil,
+			}
+		}
+	}
+	// Then loop by pipelinetask name over all the TaskRuns associated to Conditions
+	for pipelineTaskName, actualConditionTaskRuns := range conditionTaskRuns {
```

Comment:
@afrittoli `pr.Status.TaskRuns` holds an entry for the `pipelineTask` associated to a condition (could be more than one condition as well) with nil `TaskRunStatus` and one or more checks under `ConditionChecks` as in `pr.Status.TaskRuns[PipelineTaskRunName].ConditionChecks`.

Condition containers are always under `ConditionChecks` and not part of the map pr.Status.TaskRuns.

For example, for a pipelineTask with two conditions, `conditionTaskRuns` contains two containers, one for each condition and after these condition containers are created, this query returns them as valid taskRuns associated with the current pipeline: 
```c.taskRunLister.TaskRuns(pr.Namespace).List(labels.SelectorFromSet(pipelineRunLabels))```

so in the looping over those tasks, as they are conditionalChecks, line 974 is not executed and assumes the pipelineTask was not found in line 987 and creates a new taskRun name.

Hope its making sense 😉 

After the first iteration, 

PipelineRunTaskRunStatus holds map of

Reply:
Thank you so much for looking into this!

L974 is designed explicitly to only track `TaskRun` not associated to conditions, but I agree the logic is broken there. It works fine in case of orphaned conditions, but it doesn't in case of non-orphaned ones.

If the conditions are created and still running, there will be `TaskRun`s for them in the list, they can be identified as condition ones from the labels. The main `TaskRun` does not exists yet, however its name, under normal conditions, will already be in the `PipelineRun` status so that it may contain the status of the conditions:

```
taskrun-name (not yet created)
    PipelineTaskName: pipelineTaskName,
    Status:                    nil,
    ConditionChecks:  [array of condition checks],
```

Since my loops are based on what I get from the `TaskRun` list, I do not discover the existing `TaskRunName` with no real `TaskRun` under which conditions are hosted. I create a new one, and I end up having two `TaskRuns` with the same conditions - I think.

I will work on fixing this. I now separated my function in two parts, one which doesn't need the controller and takes the list of taskruns as input, so that it should be easier to write extra tests.

Reply:
I fixed the issue now by initializing `taskRunByPipelineTask` from the known PR status, and adding to it as orphaned `TaskRuns` are recovered.
I added much more test coverage, hopefully it will work ok this time.